pagefaultgames/pokerogue
# [BUGFIX] Fixed Super Fang interaction with Multi Lens

[RIP funny math formula, you will be missed.](https://www.desmos.com/calculator/wdngrksdfz)

## What are the changes the user will see?
Super Fang and company now deal correct damage with multi lens (ie 50%), split between the appropriate amount of hits linearly.
For an example of how it works: The 1st hit of super fang with 2 lenses does `(target.hp/2)*0.5=0.25*hp` damage, leaving the enemy down to 75% hp after the first hit. The remaining 2 multi lens hits each get half the difference between the leftover (75%) and the target (50%) amounts, for 12.5% damage each. 

Parental Bond works similarly, but it changes the target damage amount to 0.25 instead (the hp% the target would have left after 2 halvings).

## Why am I making these changes?
BECAUSE I COULD.

## What are the changes from a developer perspective?
Adds a check in `TargetHalfHpDamageAttr.apply()` for subsequent repeated hits that splits the damage of hits after the first one linearly to make them add up to the desired amount. The first hit is unaffected as the following strikes are intended to adjust for it appropriately.
Also adds a check in pokemon.ts to skip multiplying the move damage for the subsequent hits of such moves because they're already pre computed and I _really_ don't want to deal with reciprocals...

### Screenshots/Videos
_oh god please no I'm not spending another hour filming these just for someone to tell me to refactor them 12 hours later_

## How to test the changes?
`npm run test multi_lens`
Check the last few tests at the end.
Alternately use the following overrides:
```ts
const overrides = {
  MOVESET_OVERRIDE: Moves.SUPER_FANG,
  STARTING_LEVEL_OVERRIDE: 1000,
  STARTING_HELD_ITEMS_OVERRIDE: [{ name: "MULTI_LENS", count: 2 }],
  OPP_LEVEL_OVERRIDE: 100,
  OPP_SPECIES_OVERRIDE: Species.BLISSEY,
  OPP_MOVESET_OVERRIDE: Moves.SPLASH,
} satisfies Partial<InstanceType<typeof DefaultOverrides>>;
```
and start a new game with a Koffing and _literally any other starter without PB_.

## Checklist
- ~~If I have text, did I make it translatable and add a key in the English locale file(s)?~~
